flink-core/src/main/java/org/apache/flink/configuration/Configuration.java

```diff
@@ -850,98 +856,172 @@ private void loggingFallback(FallbackKey fallbackKey, ConfigOption<?> configOpti
+		if (o.getClass() == MemorySize.class) {
+			return (MemorySize) o;
+		} else {
+			return MemorySize.parse(o.toString());
+		}
+	}
+
+	private String convertToString(Object o) {
+		if (o.getClass() == String.class) {
+			return (String) o;
+		} else if (o.getClass() == Duration.class) {
+			Duration duration = (Duration) o;
+			return String.format("%d ns", duration.toNanos());
+		} else {
+			return o.toString();
+		}
+	}
+
+	private Integer convertToInt(Object o) {
 		if (o.getClass() == Integer.class) {
 			return (Integer) o;
-		}
-		else if (o.getClass() == Long.class) {
+		} else if (o.getClass() == Long.class) {
 			long value = (Long) o;
 			if (value <= Integer.MAX_VALUE && value >= Integer.MIN_VALUE) {
 				return (int) value;
 			} else {
-				LOG.warn("Configuration value {} overflows/underflows the integer type.", value);
-				return defaultValue;
-			}
-		}
-		else {
-			try {
-				return Integer.parseInt(o.toString());
-			}
-			catch (NumberFormatException e) {
-				LOG.warn("Configuration cannot evaluate value {} as an integer number", o);
-				return defaultValue;
+				throw new IllegalArgumentException(String.format(
+					"Configuration value %s overflows/underflows the integer type.",
+					value));
 			}
+		} else {
+			return Integer.parseInt(o.toString());
 		}
 	}
 
-	private long convertToLong(Object o, long defaultValue) {
+	private Long convertToLong(Object o) {
 		if (o.getClass() == Long.class) {
 			return (Long) o;
-		}
-		else if (o.getClass() == Integer.class) {
+		} else if (o.getClass() == Integer.class) {
 			return ((Integer) o).longValue();
-		}
-		else {
-			try {
-				return Long.parseLong(o.toString());
-			}
-			catch (NumberFormatException e) {
-				LOG.warn("Configuration cannot evaluate value " + o + " as a long integer number");
-				return defaultValue;
-			}
+		} else {
+			return Long.parseLong(o.toString());
 		}
 	}
 
-	private boolean convertToBoolean(Object o) {
+	private Boolean convertToBoolean(Object o) {
 		if (o.getClass() == Boolean.class) {
 			return (Boolean) o;
-		}
-		else {
-			return Boolean.parseBoolean(o.toString());
+		} else {
+			switch (o.toString().toUpperCase()) {
+				case "TRUE":
+					return true;
+				case "FALSE":
+					return false;
+				default:
+					throw new IllegalArgumentException(String.format(
+						"Unrecognized option for boolean: %s. Expected either true or false(case insensitive)",
+						o));
+			}
 		}
 	}
 
-	private float convertToFloat(Object o, float defaultValue) {
+	private Float convertToFloat(Object o) {
 		if (o.getClass() == Float.class) {
 			return (Float) o;
-		}
-		else if (o.getClass() == Double.class) {
+		} else if (o.getClass() == Double.class) {
 			double value = ((Double) o);
 			if (value == 0.0
-					|| (value >= Float.MIN_VALUE && value <= Float.MAX_VALUE)
-					|| (value >= -Float.MAX_VALUE && value <= -Float.MIN_VALUE)) {
+				|| (value >= Float.MIN_VALUE && value <= Float.MAX_VALUE)
```

Comment:
I feel like the previous indentation was clearer as it allowed to visually separate the check from the body of the `if`, but feel free to leave it if you prefer it this way.